src/uct/self/self_iface.c

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) UT-Battelle, LLC. 2015. ALL RIGHTS RESERVED.
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2016.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#include "self_iface.h"
+#include "self_md.h"
+#include "self_ep.h"
+
+#include <ucs/type/class.h>
+
+static ucs_config_field_t uct_self_iface_config_table[] = {
+    {"", "", NULL,
+     ucs_offsetof(uct_self_iface_config_t, super),
+     UCS_CONFIG_TYPE_TABLE(uct_iface_config_table)},
+
+    {NULL}
+};
+
+static ucs_status_t uct_self_iface_query(uct_iface_h iface, uct_iface_attr_t *attr)
+{
+    ucs_trace_func("iface=%p", iface);
+    memset(attr, 0, sizeof(uct_iface_attr_t));
+
+    attr->iface_addr_len         = sizeof(uct_self_iface_addr_t);
+    attr->device_addr_len        = sizeof(uct_self_iface_addr_t);
+    attr->ep_addr_len            = 0; //No UCT_IFACE_FLAG_CONNECT_TO_EP supported
+    attr->cap.flags              = UCT_IFACE_FLAG_CONNECT_TO_IFACE |
+                                   UCT_IFACE_FLAG_AM_SHORT         |
+                                   UCT_IFACE_FLAG_AM_CB_SYNC;
+
+    attr->cap.put.max_short      = UINT_MAX;
+    attr->cap.put.max_bcopy      = SIZE_MAX;
+    attr->cap.put.max_zcopy      = SIZE_MAX;
+
+    attr->cap.get.max_bcopy      = SIZE_MAX;
+    attr->cap.get.max_zcopy      = SIZE_MAX;
+
+    attr->cap.am.max_short       = UINT_MAX;
+    attr->cap.am.max_bcopy       = SIZE_MAX;
+    attr->cap.am.max_zcopy       = SIZE_MAX;
+    attr->cap.am.max_hdr         = 0;
+
+    attr->latency                = 10e-9; // 10 ns; TODO fix the score
+    attr->bandwidth              = 6911 * 1024.0 * 1024.0;
+    attr->overhead               = 3e-9; // 3 ns; TODO fix the score
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_self_iface_get_address(uct_iface_h iface,
+                                               uct_iface_addr_t *addr)
+{
+    ucs_trace_func("iface=%p", iface);
+    const uct_self_iface_t *local_iface = ucs_derived_of(iface, uct_self_iface_t);
+    *(uct_self_iface_addr_t*)addr = local_iface->id;
+    return UCS_OK;
+}
+
+static ucs_status_t uct_self_iface_get_device_address(uct_iface_h iface,
+                                                      uct_device_addr_t *addr)
+{
+    ucs_trace_func("iface=%p", iface);
+    return uct_self_iface_get_address(iface, (uct_iface_addr_t*) addr);
```

Comment:
Should it be replaced by ucs_empty_function_return_unsupported as function pointer?